### ISSUE TITLE
Documentation is unclear on the stored tags capability

### ISSUE THREAD

**Issue by christofmuc** [christofmuc/KnobKraft-orm#189]
Need to improve how it is possible to import with synth-specific tags.

**Comment by christofmuc** [comment]
"Mein Code liefert z.B. sowas wie:

 ['Pad’]

oder:

['SFX', 'Ambient', 'Pad’]

Da müsste doch erkannt werden, oder?

Statt dessen sehe ich im Log:

18:24:59: Warning: Synth ASM Hydrasynth has no mapping defined for stored categories. Use Categories... Edit mappings... to fix.

Die Tags, die ich mitgebe werden aber offenbar ignoriert.

Hab ich was falsch verstanden und ist die JSONC-Datei auf jeden Fall Mandatory? Oder soll ich die Standard-KnobKraft-Tags anders übergeben?"

**Comment by Andy2No** [comment]
Would it be practical to let the adaptation optionally define a numbered list of categories, and return the category number for the current patch?

E.g. the Deep Mind has a category value in each patch, presumably just a byte or smaller, and that value has a meaning which is understood by the synth and their editor software.  An adaptation could have something like:

patchCategory(); // Return the category value or -1

categoriesList(); // Return an array of strings that can be indexed with the category value

**Comment by christofmuc** [comment]
@Andy2No Ah, there is already the function 

def storedTags(self, message) -> List[str]:

Which is allowed to return a list of strings. The Adaptation Programming guide describes that additionally, you can/must supply a mapping of synth/synth-vendor specific categories to those of the user for the import to work. This is somewhat surprising in case the Adaptation produces already those that are in the database, it will still expect a mapping. 

The idea was that the adaptation should import the categories exactly as the synth vendor defined them, and then each end user can do a mapping into his or her favorite categories. Of course, it might overcomplicate things and I should check if the category by accident is exactly the same as in the database, I think.

**Comment by christofmuc** [comment]
No, bidirectional tag transfer is currently not supported. Problem was that synths support maybe only one category or two tags, and that did not look very compatible with KK. But it is possible to build it, not hard, just work. 

The external mapping file (this json or jsconc file) is also somewhat undocumented, this would need an overhaul as well.

**Comment by christofmuc** [comment]
Oh cool! 

Actually I am surprised, I was thinking you could manually rerun auto categorize and it would add tags based on the names. But it seems this is not what is implemented?

The design was always that manually assigned categories take precedence over automatically assigned categories, but I am not sure what the stored tags should fall into. 

But yes, there is this code:

```
if (result.empty()) {
	// Second step, if we have no category yet, try to detect the category from the name using the regex rule set stored in the file automatic_categories.jsonc
```

How about skipping auto categories on import, but allowing to run them later via the menu to add more categories? But that's a bit surprising, we probaby rather want both from the start.

**Comment by RadekPilich** [comment]
You're right, I'm not sure if #477 would be actually useful. Existing categories in third party patches are probably as good as it gets and for our own patches, we most likely do the assignment right on the synth before it hits KK.

**Comment by Andy2No** [comment]
@RadekPilich Cool.  Attaching it here as a zip file works.  No rush though.

**Comment by RadekPilich** [comment]
@Andy2No Here you go. 

Please a do a database backup with your currently functioning adoption in place first :)

Tweak the mapping json to fit your KK categories properly.

I do patch numbering and friendly names slightly differently then commonly done.

[mapping_categories.jsonc](https://github.com/user-attachments/files/23142075/mapping_categories.jsonc)
[Behringer Deepmind 12.py](https://github.com/user-attachments/files/23142099/Behringer.Deepmind.12.py)
